### PR TITLE
Added Windows image variants

### PR DESCRIPTION
**Pull request by sixeyed**
Signed-off-by: Elton Stoneman <elton@sixeyed.com>

Added two new Dockerfiles for building Windows variants of the NATS official image. They differ from the Linux image in two ways:

- the base image is either `microsoft/windowservercore` or `microsoft/nanoserver`
- the image installs a published release of NATS rather than building from source.

To verify NATS builds and runs correctly on Windows containers, I have used these additional images (available on Docker Hub):

- [sixeyed/nats-builder](https://hub.docker.com/r/sixeyed/nats-builder/) - which [builds and tests](https://github.com/sixeyed/dockers-windows/tree/master/nats-builder) NATS from source inside Windows containers
- [sixeyed/nats-bench](https://hub.docker.com/r/sixeyed/nats-bench/) - which is a Windows Docker image for the [NATS benchmark](https://github.com/sixeyed/dockers-windows/tree/master/nats-bench.

Functionality confirmed for the 0.9.6 release.

### CODE REVIEW

**Comment by kozlovic** [comment]
@sixeyed Elton, thanks for your contribution!

The way it works right now is that once we release a new version, we then run the script `update.sh <tag>` that will generate the corresponding `gnatsd` found in this repo. We then commit the changes and push it to this repo. When that's done, we update the [nats library](https://github.com/nats-io/official-images/blob/master/library/nats) file in the official-image repo (that we have forked). The content of the current file is this:
```
Maintainers: Derek Collison <derek@apcera.com> (@derekcollison),
             Ivan Kozlovic <ivan.kozlovic@apcera.com> (@kozlovic),
             Waldemar Salinas <wally@apcera.com> (@wallyqs)

Tags: 0.9.6, latest
GitRepo: https://github.com/nats-io/nats-docker.git
GitCommit: b2c678c7bcc6bc8d68fb33cb1f169f4b66db308d 
```
We then submit a PR, which when merged builds the docker image and pushes it to the hub.

As you can see, there is no mention of a Dockerfile, so that would not work for the windows build.

I am speculating, but I think what we would need is to specify a Directory entry in that file for each of the variant that we want. So say that in `nats-docker` we have 2 directories called `nanoserver` and `windowsservercore`, in there there will be a `Dockerfile` file (the ones that you have named with a prefix).  The nats library file then may look like this:
```
Tags: 0.9.6, latest
GitRepo: https://github.com/nats-io/nats-docker.git
GitCommit: b2c678c7bcc6bc8d68fb33cb1f169f4b66db308d 

Tags: 0.9.6-nanoserver
GitRepo: https://github.com/nats-io/nats-docker.git
GitCommit: b2c678c7bcc6bc8d68fb33cb1f169f4b66db308d 
Directory: nanoserver

Tags: 0.9.6-windowsservercore
GitRepo: https://github.com/nats-io/nats-docker.git
GitCommit: b2c678c7bcc6bc8d68fb33cb1f169f4b66db308d 
Directory: windowsservercore
```
But I honestly don't know if this is the way to do it. Maybe we need to have different library files and so different nats-docker repos?

Also, it means that NATS team would be the official maintainer for that. I thought that you would have your own repo and submit a PR to official-images on your own and we would simply give our blessing (so that people at Docker approve the official image from someone not part of the upstream's organization).

We will discuss this internally and get back to you quickly. Thanks again!

**Comment by sixeyed** [comment]
@kozlovic thanks for checking - yes you're right that I should have different directories for the different base images. That's done now.

On the maintainer front I don't think an official image can use different source repos for different image variants, but I'm happy to add me in a `MAINTAINER` instruction in the Windows Dockerfiles.

**Comment by kozlovic** [comment]
Good! Do you think we need a copy of `gnatsd.conf` in each directory or can we refer to `../../gnatsd.conf` in the Dockerfile? (I am not sure if we can do that).

**Comment by sixeyed** [comment]
No, can't do that Ivan - Docker uses the folder with the Dockerfile as the
build context, and you can't refer to files in parent directories.

On the plus side it means we could have different config settings if we
needed to.

**Comment by kozlovic** [comment]
Ok thanks for checking, I figured it would be the case. We just need to remember to copy the from the root dir to each windows directory if one day we change the content of `gnatsd.conf`.

Also, I would have prefer that the script `update.sh` generates the windows Dockerfile (or say have a template in which the version is passed and generates the actual Dockerfile files) so we don't have to manually edit each of the Dockerfile to bump the version, but `update.sh` works also for branches, not only tags, so I guess it is fine for now.

So do you know if what I suggested about having different entries with different tags/directories in the `nats` library of the docker official-images repo would work? I think it would but...

We still need to discuss this internally and we will get back to you.

**Comment by sixeyed** [comment]
Yes, adding it to the script should be straightforward. The other libraries with multiple variants do something similar. 

On the library metadata, the directory path is how the Dockerfile gets found - like with [golang](https://github.com/nats-io/official-images/blob/master/library/golang). The new manifest would be:

```
Tags: 0.9.6, latest
GitRepo: https://github.com/nats-io/nats-docker.git
GitCommit: b2c678c7bcc6bc8d68fb33cb1f169f4b66db308d 

Tags: 0.9.6-nanoserver, nanoserver
GitRepo: https://github.com/nats-io/nats-docker.git
GitCommit: [COMMIT]
Directory: windows/nanoserver
Constraints: nanoserver

Tags: 0.9.6-windowsservercore, windowsservercore
GitRepo: https://github.com/nats-io/nats-docker.git
GitCommit: [COMMIT]
Directory: windows/windowsservercore
Constraints: windowsservercore
```
Looking at that, we could have the Windows Dockerfiles in a separate Git repo. If you prefer I can set up `sixeyed/nats-docker-windows` and maintain it separately from this repo.

**Comment by kozlovic** [comment]
Ok, got it (we can actually simply by moving the common GitRepo and GitCommit to the global metadata section so not to have to repeat them).

About creating a different repo, I am not sure. At this point in time, if we were to produce a new version of NATS, I would not be able to test the Windows docker images, so then it means that you would responsible to build and check. That would make sense to have 2 different repos since you would be the sole maintainer.
But it means that when issuing a new release both of us would have to coordinate and so that you update the version number, build, test and push to your repo. With the new commit sha, only then would I be able to submit the PR for the official image.. unless we also have a different library, say nats-windows.

**Comment by sixeyed** [comment]
If we're coordinating to release the image, then I guess if there's a lag then it's better if the Windows version is delayed, rather than the whole release? In that case a separate repo would make more sense. Happy for that to live in my account or in `nats-io`.

**Comment by kozlovic** [comment]
Yes, but if we are concerned about delays, then not only you would need a different repo for the Docker files (what you have in this PR), but it means a different library file in official-images. Do we want to go there? What do you think? I will also discuss with couple of people here.

**Comment by sixeyed** [comment]
Personally, no. It's the same NATS product running on a different platform, so to me it's one image repository with several variants - `nats`, `nats:nanoserver` etc. I don't think a release delay will actually be an issue in practice. If it does become a problem we can update the script to generate the Windows Dockerfiles at the same time as the Linux one.

**Comment by kozlovic** [comment]
@sixeyed After discussion, we think that it would be ok to merge this PR in this repo and then submit a PR for official-images with the windows variants as discussed above.

Now, looking at other official images that provide windows variants, I am having a doubt and would like your opinion: are they (docker people) building the image for windows variants, and pushing those to the hub? I see some outputs for PRs of products having windows variant that seem to indicate that build/test is skipped because of the constraints (not running on nano or windowsservercore): for instance this PR: https://github.com/docker-library/official-images/pull/2376, where you can see the build comment entry:
```
$ bashbrew build hello-world:latest
Using bashbrew/cache:c2f2c94c73d3411768c20defee455f47200efff60b1c2bb224019d0f48023cf7 (hello-world:latest)
Tagging hello-world:latest

$ test/run.sh hello-world:latest
testing hello-world:latest
	'cve-2014--shellshock' [1/2]...passed
	'override-cmd' [2/2]...image does not appear to contain "echo" -- assuming single-binary image
passed


$ bashbrew build hello-world:nanoserver
skipping "hello-world:nanoserver" (due to unsatisfactory constraints ["nanoserver"])

$ test/run.sh hello-world:nanoserver
testing hello-world:nanoserver
	image does not exist!
```
If you have some insight, I would appreciate your comments.

**Comment by kozlovic** [comment]
@sixeyed Any thoughts on my last comment? Thanks!

**Comment by sixeyed** [comment]
@tianon can you comment on this? We have working Dockerfiles for Windows variants that the upstream project are happy to merge, but can we just change the manifest for `official-images` and send a PR, or do we need an additional step to build the Windows images? LMK if I can help.

**Comment by sixeyed** [comment]
@friism - can I ask for your input on this ^^

**Comment by friism** [comment]
@tianon @yosifkit can you help with guidance?

**Comment by tianon** [comment]
@kozlovic these look pretty good to me -- my only thought/recommendation would be to ensure that they've got as much parity with the non-Windows variants as is possible/sane (ie, similar configuration, similar `VOLUME` declarations, etc)

Regarding the official-images, we usually do a separate build test, and do the build/push from a separate server at build time.  Just make sure that you use the appropriate `nanoserver` and `windowsservercore` values in the `Constraints` field of your library file (as in Redis, for example: https://github.com/docker-library/official-images/blob/9dc0afe7b04d8f5ca9de992cc5a243a0a353f180/library/redis#L41-L49)
```
Tags: 3.2.6-windowsservercore, 3.2-windowsservercore, 3-windowsservercore, windowsservercore
GitCommit: 709e6a8df205f73afcc6f6257cab104175de1f5f
Directory: 3.2/windows/windowsservercore
Constraints: windowsservercore

Tags: 3.2.6-nanoserver, 3.2-nanoserver, 3-nanoserver, nanoserver
GitCommit: 2daf4fd7d858c8f2ad9694c4f460b8c6a7f782aa
Directory: 3.2/windows/nanoserver
Constraints: nanoserver
```

**Comment by kozlovic** [comment]
@tianon Thanks! By the way, if we submit a new PR to official-images to include the Windows variants, do we need to have a new release (new tag) or can we add to the existing tag? In other words, will your build process be triggered if it does not detect a new tag?

**Comment by tianon** [comment]
Our build process is triggered by PRs (and changes to that `GitCommit` value, which is why we prefer an explicit commit to a tag), so feel free to fire away whenever you're ready! :+1:

File: windows/nanoserver/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    Remove-Item gnatsd.zip
+
+WORKDIR c:/gnatsd
+COPY gnatsd.conf gnatsd.conf
+
+# Expose client, management, and cluster ports
+EXPOSE 4222 8222 6222
+
+ENTRYPOINT ["gnatsd"]
+CMD ["-c", "gnatsd.conf"]
```
Comment (friism) [inline windows/nanoserver/Dockerfile]: missing newline at EOF

File: windows/windowsservercore/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    Remove-Item gnatsd.zip
+
+WORKDIR c:/gnatsd
+COPY gnatsd.conf gnatsd.conf
+
+# Expose client, management, and cluster ports
+EXPOSE 4222 8222 6222
+
+ENTRYPOINT ["gnatsd"]
+CMD ["-c", "gnatsd.conf"]
```
Comment (friism) [inline windows/windowsservercore/Dockerfile]: missing newline

**Comment by kozlovic** [comment]
The new line is also missing from the original Dockerfile. @sixeyed if you change the Windows one, would you mind changing the original too?
@friism I already pointed that a few comments back, and here was his response:
```
Good! Do you think we need a copy of gnatsd.conf in each directory or can we refer to ../../gnatsd.conf in the Dockerfile? (I am not sure if we can do that).
     
 sixeyed commented 13 days ago
No, can't do that Ivan - Docker uses the folder with the Dockerfile as the
build context, and you can't refer to files in parent directories.

On the plus side it means we could have different config settings if we
needed to.
```

**Comment by yosifkit** [comment]
Now that you might have multiple versions, for ease of generating the `library/nats` file, you might want to look at some of the repos in `docker-library` like `redis` and see the [`generate-stackbrew-library.sh`](https://github.com/docker-library/redis/blob/12bbc958e61d7d59a92042e60391fdb9188bb8aa/generate-stackbrew-library.sh) script.   We use it like this: 
```console
$ # once the current checkout is up to date with github
$ ./generate-stackbrew-library.sh > ../official-images/library/redis
$ # then go commit it and make a PR in the official-images
```
If you want help adapting something similar for nats, let me know.

**Comment by friism** [comment]
ok, nevermind - sorry, I should have checked the conversation

**Comment by kozlovic** [comment]
@yosifkit Do we really need that for nats? It is likely at after this first PR, the commit will be the same for all variants (I was even thinking of bringing the GitCommit with the GitRepo up in the meta section since it would apply to all variants)?

**Comment by tianon** [comment]
@kozlovic that sounds fine :smile: :+1:

(I think @yosifkit was just noting that's a pattern we've had a lot of success with in case you found it helpful here too :+1:)

**Comment by kozlovic** [comment]
@tianon Thanks, yes I saw that from other projects and will look into it.

**Comment by sixeyed** [comment]
Updated from @friism's review - should be good to go now.

**Pull request merged**